cocos/core/scene-graph/sorting-manager.ts

```diff
@@ -0,0 +1,137 @@
+
+export class SortingManager {
+    public static getSortingPriority (layer = 0, order = 0): number {
+        return (((layer + (1 << 15)) << 16) | (order + (1 << 15))) >>> 0;
+    }
+
+    public static getSortingLayerIndex (layerID = 0): number {
+        let index = 0;
+        if (this.indexMap.has(layerID)) {
+            index = this.indexMap.get(layerID)!;
+        } else {
+            console.error('invalid ID');
+        }
+        return index;
+    }
+
+    public static getSortingIndexFromName (name: string): number {
+        const id = this.getSortingIDFromName(name);
+        return this.getSortingLayerIndex(id);
+    }
+
+    // ID To Name
+    public static getSortingLayerName (layerID = 0): string {
+        let name = '';
+        if (this.nameMap.has(layerID)) {
+            name = this.nameMap.get(layerID)!;
+        } else {
+            console.error('invalid ID');
+        }
+        return name;
+    }
+
+    // Name To ID
+    public static getSortingIDFromName (name: string): number {
+        const count = this.nameMap.size;
+        const keyIterator = this.nameMap.keys();
+        let key = 0;
+        for (let i = 0; i < count; i++) {
+            key = keyIterator.next().value;
+            if (this.nameMap.get(key) === name) return key;
+        }
+        console.warn('invalid name');
+        return 0;
+    }
+
+    public static idIsValid (id: number): boolean {
+        // check valid
+        if (this.indexMap.has(id)) {
+            return true;
+        } else {
+            console.error('invalid ID');
+            return false;
+        }
+    }
+
+    // Editor Function
+
+    public static addSortingLayer () {
+        this.ID++;
+        if (this.ID > 65535) {
+            console.error('too many layers');
+        }
+        this.nameMap.set(this.ID, `New Layer${this.ID}`);
+        this.indexMap.set(this.ID, this.ID);
+    }
+
+    public static removeSortingLayer (layerID: number) {
+        if (!this.idIsValid(layerID)) return;
+
+        if (this.nameMap.has(layerID) && this.indexMap.has(layerID)) {
+            // Todo: need update all component used this layer
+            // set the value to 0
+            this.nameMap.delete(layerID);
+            this.indexMap.delete(layerID);
+        }
+    }
+
+    public static renameSortingLayer (layerID, layerName) {
+        if (this.nameMap.has(layerID)) {
+            console.warn('Invalid layer id.');
+        } else {
+            this.nameMap.set(layerID, layerName);
+        }
+    }
+
+    public static changeIndex () {
+        // Todo: need update index map
+        // Todo: need update all component sorting priority in active scene
+    }
+
+    private static defaultIndex = 0;
+    private static defaultName = 'default';
+    private static nameMap = new Map<number, string>();
+    private static indexMap = new Map<number, number>();
+    private static ID = 0;
+
+    public static initLayerFromSettings (layerID, layerName, layerIndex) {
```

Comment:
There is no settings in the input parameter, I think it's the same ability as `addLayer`, we can have 

```
public static addLayer(layer, layerName, layerIndex)
```

Do some assert and check in the function is enough, don't over complicate the usage here

Reply:
It will result in ID needing to be managed externally, rather than assigned internally.